tests/testthat/test-missing_data.R

```diff
@@ -25,12 +25,68 @@ test_that("defMiss produces correct output", {
 # genMiss ----
 test_that("genMiss works", {
   # not repeated
+  def1 <- defData(varname = "m", dist = "binary", formula = .5)
+  def1 <- defData(def1, "u", dist = "binary", formula = .5)
+  def1 <- defData(def1, "x1", dist="normal", formula = "20*m + 20*u", variance = 2)
+  def1 <- defData(def1, "x2", dist="normal", formula = "20*m + 20*u", variance = 2)
+  def1 <- defData(def1, "x3", dist="normal", formula = "20*m + 20*u", variance = 2)
   
-  # repeated
+  dtAct <- genData(1000, def1)
+  
+  hardProbForm <- runif(1)
+  form1 <- ".05 + m * 0.25"
+  form2 <- ".05 + u * 0.25"
+  
+  defM <- defMiss(varname = "x1", formula = hardProbForm, logit.link = FALSE)
+  defM <- defMiss(defM, varname = "x2", formula = form1, logit.link = FALSE)
+  defM <- defMiss(defM, varname = "x3", formula = form2, logit.link = FALSE)
+  defM <- defMiss(defM, varname = "u", formula = 1, logit.link = FALSE) # not observed
+  
+  missMat <- genMiss(dtAct, defM, idvars = "id")
+  
+  ## check probabilistic column has correct distribution
+  hardProbAct <- sum(missMat[, x1]) / nrow(missMat)
+  hPDiff <- abs(hardProbAct - hardProbForm)
+  
+  expect_true(hPDiff < 0.02)
+  
+  ## check dependent column 1
+  mAvg <- sum(dtAct[, m]) / nrow(dtAct)
+  m <- mAvg
+  col1Exp <- eval(parse(text = as.character(form1)))
+  col1Act <- sum(missMat[, x2]) / nrow(missMat)
+  col1Diff <- abs(col1Act - col1Exp)
+  
+  expect_true(col1Diff < 0.02)
+  
+  ## check dependent column 2
+  uAvg <- sum(dtAct[, u]) / nrow(dtAct)
+  u <- uAvg
+  col2Exp <- eval(parse(text = as.character(form2)))
+  col2Act <- sum(missMat[, x3]) / nrow(missMat)
+  col2Diff <- abs(col2Act - col2Exp)
+  
+  expect_true(col2Diff < 0.02)
+  
+  ## check empty column
+  expect_true(all(missMat[, u] == 1))
```

Comment:
I realize that this is all pretty hard coded but I wanted to get something down so I could get some feedback from you. I feel like the concepts are there but I don't know that this is really the best way to go about testing this.

Also, I'm not understanding how genMiss works when repeated = TRUE very well, and so I'm having trouble coming up with testing ideas.

Reply:
The check for `u` in `missMat` is good. Also could check for `x2` and `x3` - missingness will depend on the actual value of `m` but not `u`. Also maybe want to confirm that `m` is all 0's.

I'll explain the `repeated = TRUE` option.